.github/workflows/sync-gitster-git.yml

```diff
@@ -0,0 +1,120 @@
+name: sync-gitster-git-branches
+
+on:
+  schedule:
+    - cron: '17 6 * * *'
+  workflow_dispatch:
+
+env:
+  SOURCE_REPOSITORY: gitster/git
+  TARGET_REPOSITORY: gitgitgadget/git
+
+jobs:
+  sync-gitster-git-branches:
+    runs-on: ubuntu-latest
+    steps:
+      - name: check which refs need to be synchronized
+        uses: actions/github-script@v6
+        id: check
+        with:
+          script: |
+            const sleep = async (milliseconds) => {
+              return new Promise(resolve => setTimeout(resolve, milliseconds))
+            }
+
+            const getRefs = async (repository) => {
+              for (let i = 1; i <= 10; i++) {
+                try {
+                  const [owner, repo] = repository.split('/')
+                  const { data } = await github.rest.git.listMatchingRefs({
+                    owner,
+                    repo,
+                    ref: 'heads/'
+                  })
+                  return data
+                    .filter(e => e.ref.match(/^refs\/heads\/(maint-\d|[a-z][a-z]\/)/))
+                    .sort((a, b) => a.ref.localeCompare(b.ref))
+                } catch (e) {
+                  if (e?.status !== 502) throw e
+                }
+
+                const seconds = i * i + 15 * Math.random()
+                core.info(`Encountered a Server Error; retrying in ${seconds} seconds`)
+                await sleep(1000 * seconds)
+              }
+            }
+
+            const sourceRefs = await getRefs(process.env.SOURCE_REPOSITORY)
+            const targetRefs = await getRefs(process.env.TARGET_REPOSITORY)
+
+            const refspecs = []
+            const toFetch = new Set()
+            for (let i = 0, j = 0; i < sourceRefs.length || j < targetRefs.length; ) {
+              const c =  i >= sourceRefs.length
+                ? +1
+                : j >= targetRefs.length
+                  ? -1
+                  : sourceRefs[i].ref.localeCompare(targetRefs[j].ref)
+              if (c > 0) {
+                // no source ref => delete target ref
+                refspecs.push(`:${targetRefs[j++].ref}`)
+              } else if (c < 0) {
+                // no corresponding target ref yet => push source ref (new)
+                const sha = sourceRefs[i].object.sha
+                toFetch.add(sha)
+                refspecs.push(`${sha}:${sourceRefs[i++].ref}`)
+              } else if (c === 0 && sourceRefs[i].object.sha !== targetRefs[j].object.sha) {
+                // target ref needs updating
+                const sha = sourceRefs[i].object.sha
+                toFetch.add(sha)
+                refspecs.push(`+${sha}:${sourceRefs[i++].ref}`)
+                j++
+              } else {
+                i++
+                j++
+              }
```

Comment:
Maybe a little clearer on the increments?

```
              } else { // they are a match
               if (sourceRefs[i].object.sha !== targetRefs[j].object.sha) {
                  // target ref needs updating
                  const sha = sourceRefs[i].object.sha
                  toFetch.add(sha)
                  refspecs.push(`+${sha}:${sourceRefs[i].ref}`)
                } 
                i++
                j++
              }
```

Reply:
Good point!